pkg/cmd/roachprod/grafana/annotations.go

```diff
@@ -0,0 +1,126 @@
+		}
+		ts, err := idtoken.NewTokenSource(ctx, grafanaAudience, idtoken.WithCredentialsJSON([]byte(grafanaKey)))
+		if err != nil {
+			return nil, errors.Wrap(err, "Error creating GCS oauth token source from specified credential")
+		}
+		token, err := ts.Token()
+		if err != nil {
+			return nil, errors.Wrap(err, "Error getting token")
+		}
+
+		headers["Authorization"] = fmt.Sprintf("Bearer %s", token.AccessToken)
+	}
+
+	headers[httputil.ContentTypeHeader] = httputil.JSONContentType
+	cfg := &grafclient.TransportConfig{
+		Host:        host,
+		BasePath:    "/api",
+		Schemes:     []string{scheme},
+		HTTPHeaders: headers,
+	}
+
+	return grafclient.NewHTTPClientWithConfig(strfmt.Default, cfg), nil
+}
+
+// AddGrafanaAnnotation creates an HTTP client and sends a POST request to the
+// specified host name to create an annotation. If successful, it returns
+// the result.
+func AddGrafanaAnnotation(
+	ctx context.Context, host string, secure bool, req AddAnnotationRequest,
+) (string, error) {
+	// Cleanup the host name. The Grafana OpenAPI already appends the scheme
+	// and leading slash.
+	host, _ = strings.CutPrefix(host, "https://")
+	host, _ = strings.CutPrefix(host, "http://")
+	host = strings.Trim(host, "/")
+
+	httpClient, err := newGrafanaClient(ctx, host, secure)
+	if err != nil {
+		return "", err
+	}
+
+	body := &models.PostAnnotationsCmd{
+		Text:         &req.Text,
+		Tags:         req.Tags,
+		DashboardUID: req.DashboardUID,
+		Time:         req.StartTime,
+		TimeEnd:      req.EndTime,
+	}
+
+	res, err := httpClient.Annotations.PostAnnotation(body)
+	if err != nil {
+		emptyResponseBodyErr := fmt.Errorf("&{ <nil> } (*models.ErrorResponseBody) is not supported by the TextConsumer, can be resolved by supporting TextUnmarshaler interface")
+		// The Grafana OpenAPI does not correctly handle empty responses, i.e. renegotiating
+		// the content-type dynamically if there is an error with no body. This leads to a
+		// cryptic `(*models.ErrorResponseBody) is not supported by the TextConsumer` error,
+		// but is most likely a 401 authentication issue so give the caller a hint.
+		if errors.Is(err, emptyResponseBodyErr) {
+			return "", errors.Wrap(err, "AddGrafanaAnnotation failed - HINT: if issue is not transient, check that you are authenticating correctly")
+		}
+		return "", err
+	}
+	if !res.IsSuccess() {
+		return "", errors.Newf("AddGrafanaAnnotation failed with: %s", res)
+	}
+
+	return res.String(), nil
+}
+
+// AddAnnotationRequest are the fields used to create a Grafana annotation
+// post request. Only Text is a required field.
+type AddAnnotationRequest struct {
+	Text         string
+	DashboardUID string
+	Tags         []string
+	// The Grafana API expects the start and end time to be in epoch millisecond time.
+	// If empty, creates an annotation at the current time. If both StartTime and
+	// EndTime are specified, creates an annotation over a time range.
```

Comment:
Time range being `[StartTime, EndTime)`, or is `EndTime` inclusive? Let's be precise in the comment since these things tend to induce off-by-one errors.

Reply:
Grafana doesn't specify but some testing shows it's inclusive. I think this would only matter if someone was trying to query for annotations over a range of time, currently not supported through roachprod, but I added a comment anyway to be clear.